pkg/daemon/server/service/pipeline_watermark_query.go

```diff
@@ -0,0 +1,107 @@
+// Package service is built for querying metadata and to expose it over daemon service.
+package service
+
+import (
+	"context"
+	"fmt"
+	"github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	"github.com/numaproj/numaflow/pkg/apis/proto/daemon"
+	"github.com/numaproj/numaflow/pkg/isbsvc"
+	"github.com/numaproj/numaflow/pkg/isbsvc/clients"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"github.com/numaproj/numaflow/pkg/watermark/fetch"
+	"github.com/numaproj/numaflow/pkg/watermark/generic"
+	"github.com/numaproj/numaflow/pkg/watermark/store/jetstream"
+	"go.uber.org/zap"
+	"time"
+)
+
+// watermarkFetchers used to store watermark metadata for propagation
+type watermarkFetchers struct {
+	fetchMap           map[string]fetch.Fetcher
+	isWatermarkEnabled bool
+}
+
+// newVertexWatermarkFetcher creates a new instance of watermarkFetchers. This is used to populate a map of vertices to
+// corresponding fetchers. These fetchers are tied to the incoming edge buffer of the current vertex (Vn), and read the
+// watermark propagated by the vertex (Vn-1). As each vertex has one incoming edge, for the input vertex we read the source
+// data buffer.
+func newVertexWatermarkFetcher(pipeline *v1alpha1.Pipeline) *watermarkFetchers {
+	ctx := context.Background()
+	log := logging.FromContext(ctx)
+	var wmFetcher = new(watermarkFetchers)
+	var fromBufferName string
+
+	wmFetcher.isWatermarkEnabled = pipeline.Spec.Watermark.Propagate
+	if !wmFetcher.isWatermarkEnabled {
+		return wmFetcher
+	}
+
+	vertexWmMap := make(map[string]fetch.Fetcher)
+	pipelineName := pipeline.Name
+
+	for _, vertex := range pipeline.Spec.Vertices {
+		// TODO: Checking if Vertex is source
+		if vertex.Source != nil {
+			fromBufferName = v1alpha1.GenerateSourceBufferName(pipeline.Namespace, pipelineName, vertex.Name)
+		} else {
+			// Currently we support only one incoming edge
+			edge := pipeline.GetFromEdges(vertex.Name)[0]
+			fromBufferName = v1alpha1.GenerateEdgeBufferName(pipeline.Namespace, pipelineName, edge.From, edge.To)
+		}
+		hbBucket := isbsvc.JetStreamProcessorBucket(pipelineName, fromBufferName)
```

Comment:
It could be ISB svc other than Jetstream.

Reply:
if ISB Svc is not jetstream, watermark cannot be propagated. We will not execute this part of the code. Do you have a better suggestion?

Reply:
`otWatch` and `hbWatch` have to be 2 passed in interfaces, which are created in `daemonServer.Run()`,  at the place `switch ds.isbSvcType {`, and then all the way down from `newGRPCServer()` and `NewPipelineMetricsQueryService`.